Reader volume

gdpelican opened the pull request:
Not for merging yet, but a solid first pass at converting the idea of 'following' to 'volume'

robguthrie commented:
Looks like a great start @gdpelican I'm super pleased with the direction it's taking.

robguthrie commented:
you might need to rebase off base though.. looks like my merge sqashing as confused your branch and you're trying to put in some old code.

robguthrie commented on app/models/discussion.rb:
```diff
@@ -77,7 +77,7 @@ def followers
       joins("LEFT OUTER JOIN discussion_readers dr ON (dr.user_id = users.id AND dr.discussion_id = #{id})").
       joins("LEFT OUTER JOIN memberships m ON (m.user_id = users.id AND m.group_id = #{group_id})").
-      where('dr.following = TRUE OR (dr.following IS NULL AND m.following_by_default = TRUE)')
+      where('dr.volume = :email OR (dr.volume IS NULL AND m.volume = :email)', { email: DiscussionReader.volumes[:email] })
```
Are you super sure you want to call membership.volume that? I feel like it might make it clearer if its default_volume on membership

gdpelican commented:
Yeah, trying to patch that stuff up now.

On Thu, Feb 26, 2015 at 9:49 AM, Robert Guthrie <notifications@github.com>
wrote:

> you might need to rebase off base though.. looks like my merge sqashing as
> confused your branch and you're trying to put in some old code.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/loomio/loomio/pull/1950#issuecomment-76053484>.
>

gdpelican commented:
@robguthrie this is ready for a smoke testing.

gdpelican commented:
@robguthrie I added serializing out permitted params on here, which seems to work swimmingly. Also fixed the create new discussion bug by filtering out discussions without an id from the inbox.

(The bug was that we were calling `activeProposal()` -> `proposals()` on a discussion which had not been created yet, which means it's `setupViews()` hadn't been called yet, which means b0rkdom.)

gdpelican commented on app/models/discussion_reader.rb:
```diff
@@ -19,20 +20,14 @@ def self.for(user: nil, discussion: nil)
-  def follow!
-    update_attribute(:following, true)
+  def set_volume_as_required!
+    if user.email_on_participation?
```
Nested unless / if :(

```
if user.email_on_participation? && !volume_is_email?
  set_volume! :email
end
```

gdpelican commented on app/models/motion.rb:
```diff
@@ -226,11 +214,11 @@ def update_vote_counts!
 
   def group_members_without_motion_author
-    group.users.where(User.arel_table[:id].not_eq(author.id))
+    group.members.without(author)
```
These are cool.

gdpelican commented on app/models/user.rb:
```diff
@@ -114,17 +114,35 @@ class User < ActiveRecord::Base
+    if users.size > 0
+      where('users.id NOT IN (?)', users)
+    else
+      all
```
I think you want `scoped` here, so it'll handle default scopes correctly.

robguthrie replied:
scoped was deprecated in rails 4.0

gdpelican commented on app/queries/users_by_thread_volume_query.rb:
```diff
@@ -0,0 +1,22 @@
+class UsersByThreadVolumeQuery
```
Any reason we're ditching the `Queries::UsersByThreadVolume` idiom for these?

robguthrie replied:
rails forces its naming conventions hard. moving queries into app means you namespace on 2nd level depth. So Queries:: is basically wrong.. and it has to live in app/extras/queries just to work as is.

gdpelican commented on app/queries/users_by_thread_volume_query.rb:
```diff
@@ -0,0 +1,22 @@
+  def self.volume_and_discussion(volume, discussion)
+    User.
+      active.
+      joins("LEFT OUTER JOIN discussion_readers dr ON (dr.user_id = users.id AND dr.discussion_id = #{discussion.id})").
```
Probably fine, but just to be more safest, `... dr.discussion_id = ?", discussion.id`

robguthrie replied:
you can't do that in a join :(

gdpelican commented on app/queries/users_to_email_query.rb:
```diff
@@ -0,0 +1,40 @@
+  def self.new_comment(comment)
+    discussion = comment.discussion
+
+    UsersByThreadVolumeQuery.email(discussion).
```
`.email(comment.discussion)` ?

gdpelican commented on app/views/thread_mailer/_footer.html.haml:
```diff
@@ -11,10 +11,10 @@
+  - else
     - follow_url = email_actions_follow_discussion_url(discussion_id: @discussion.id, unsubscribe_token: @recipient.unsubscribe_token)
-    = t :'email.thread_mailer.follow_discussion_html', follow_url: follow_url
+    = t :'email.thread_mailer.turn_on_notifications_html', follow_url: follow_url
```
follow_url -> url here as well?

robguthrie replied:
oh, thanks
